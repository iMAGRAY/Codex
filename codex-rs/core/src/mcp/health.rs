use std::time::SystemTime;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum HealthStatus {
    Unknown,
    Passing,
    Failing,
}

#[derive(Debug, Clone)]
pub struct HealthReport {
    pub status: HealthStatus,
    pub last_checked_at: Option<SystemTime>,
    pub notes: Option<String>,
}

impl HealthReport {
    pub fn new(status: HealthStatus) -> Self {
        Self {
            status,
            last_checked_at: None,
            notes: None,
        }
    }

    pub fn with_notes(mut self, notes: impl Into<String>) -> Self {
        self.notes = Some(notes.into());
        self
    }

    pub fn unknown() -> Self {
        Self::new(HealthStatus::Unknown).with_notes("health checks not implemented yet")
    }
}
