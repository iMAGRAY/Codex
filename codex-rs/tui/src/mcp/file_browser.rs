use std::cell::Cell;
use std::cmp::Ordering;
use std::fs;
use std::path::Path;
use std::path::PathBuf;

use ratatui::buffer::Buffer;
use ratatui::layout::Constraint;
use ratatui::layout::Layout;
use ratatui::layout::Rect;
use ratatui::style::Modifier;
use ratatui::style::Style;
use ratatui::style::Stylize;
use ratatui::text::Line;
use ratatui::text::Span;
use ratatui::widgets::Block;
use ratatui::widgets::Borders;
use ratatui::widgets::Cell as TableCell;
use ratatui::widgets::Paragraph;
use ratatui::widgets::Row;
use ratatui::widgets::StatefulWidget;
use ratatui::widgets::Table;
use ratatui::widgets::TableState;
use ratatui::widgets::Widget;
use ratatui::widgets::Wrap;

use crate::bottom_pane::ScrollState;

#[derive(Clone, Debug)]
pub(crate) struct FileBrowserEntry {
    label: String,
    summary: Option<String>,
    path: PathBuf,
    kind: FileBrowserEntryKind,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub(crate) enum FileBrowserEntryKind {
    Current,
    Parent,
    Directory,
    Manifest,
}

pub(crate) enum FileBrowserOutcome {
    None,
    Selected(String),
    Cancelled,
}

pub(crate) struct FileBrowser {
    workspace_root: PathBuf,
    current_dir: PathBuf,
    entries: Vec<FileBrowserEntry>,
    scroll: ScrollState,
    error: Option<String>,
    visible_rows: Cell<usize>,
}

impl FileBrowser {
    pub(crate) fn new(start: PathBuf, workspace_root: PathBuf) -> Self {
        let normalized_root = workspace_root;
        let mut browser = Self {
            workspace_root: normalized_root.clone(),
            current_dir: normalize_start(&start, &normalized_root),
            entries: Vec::new(),
            scroll: ScrollState::new(),
            error: None,
            visible_rows: Cell::new(1),
        };
        browser.refresh_entries();
        browser
    }

    pub(crate) fn render(&self, area: Rect, buf: &mut Buffer, title: &str) {
        let help_height = 4;
        let header_height = 3;
        let layout = Layout::vertical([
            Constraint::Length(header_height),
            Constraint::Min(area.height.saturating_sub(header_height + help_height)),
            Constraint::Length(help_height),
        ])
        .split(area);

        let header_lines = vec![
            Line::from(vec![
                "Browse from: ".dim(),
                relative_display(&self.current_dir, &self.workspace_root).into(),
            ]),
            Line::from(vec![
                "Root: ".dim(),
                self.workspace_root.to_string_lossy().into_owned().into(),
            ]),
        ];
        let header = Paragraph::new(header_lines)
            .block(Block::default().borders(Borders::ALL).title(title))
            .wrap(Wrap { trim: true });
        Widget::render(header, layout[0], buf);

        self.render_table(layout[1], buf);

        let mut help_lines = Vec::new();
        if let Some(err) = &self.error {
            help_lines.push(err.as_str().red().into());
        }
        help_lines.push(Line::from(vec![
            "Enter".cyan(),
            " select  ".dim(),
            "‚Üí/l".cyan(),
            " open  ".dim(),
            "‚Üê/h/Backspace".cyan(),
            " up  ".dim(),
            "space".cyan(),
            " select highlighted  ".dim(),
            "Esc".cyan(),
            " cancel".dim(),
        ]));
        let help = Paragraph::new(help_lines)
            .block(Block::default().borders(Borders::ALL).title("Controls"))
            .wrap(Wrap { trim: true });
        Widget::render(help, layout[2], buf);
    }

    fn render_table(&self, area: Rect, buf: &mut Buffer) {
        let rows: Vec<Row> = self
            .entries
            .iter()
            .map(|entry| {
                let style = match entry.kind {
                    FileBrowserEntryKind::Current => Style::default().add_modifier(Modifier::BOLD),
                    FileBrowserEntryKind::Manifest => {
                        Style::default().fg(ratatui::style::Color::Yellow)
                    }
                    _ => Style::default(),
                };
                Row::new(vec![
                    TableCell::from(Span::styled(entry.label.clone(), style)),
                    TableCell::from(Span::styled(
                        entry.summary.clone().unwrap_or_default(),
                        style,
                    )),
                ])
            })
            .collect();

        let mut state = TableState::default();
        state.select(self.scroll.selected_idx);
        *state.offset_mut() = self.scroll.scroll_top;

        let table = Table::new(
            rows,
            [Constraint::Percentage(45), Constraint::Percentage(55)],
        )
        .header(Row::new(vec!["Entry".bold(), "Summary".bold()]))
        .block(Block::default().borders(Borders::ALL).title("Workspace"))
        .row_highlight_style(Style::default().add_modifier(Modifier::REVERSED));
        StatefulWidget::render(table, area, buf, &mut state);

        let visible_rows = area.height.saturating_sub(3) as usize;
        self.visible_rows.set(visible_rows.max(1));
    }

    pub(crate) fn handle_key(&mut self, key: crossterm::event::KeyEvent) -> FileBrowserOutcome {
        use crossterm::event::KeyCode;
        self.error = None;
        match key.code {
            KeyCode::Esc | KeyCode::Char('q') => FileBrowserOutcome::Cancelled,
            KeyCode::Up => {
                self.scroll.move_up_wrap(self.entries.len());
                self.scroll
                    .ensure_visible(self.entries.len(), self.visible_rows());
                FileBrowserOutcome::None
            }
            KeyCode::Down => {
                self.scroll.move_down_wrap(self.entries.len());
                self.scroll
                    .ensure_visible(self.entries.len(), self.visible_rows());
                FileBrowserOutcome::None
            }
            KeyCode::PageUp => {
                let steps = self.visible_rows().saturating_sub(1);
                for _ in 0..steps {
                    self.scroll.move_up_wrap(self.entries.len());
                }
                self.scroll
                    .ensure_visible(self.entries.len(), self.visible_rows());
                FileBrowserOutcome::None
            }
            KeyCode::PageDown => {
                let steps = self.visible_rows().saturating_sub(1);
                for _ in 0..steps {
                    self.scroll.move_down_wrap(self.entries.len());
                }
                self.scroll
                    .ensure_visible(self.entries.len(), self.visible_rows());
                FileBrowserOutcome::None
            }
            KeyCode::Enter | KeyCode::Char(' ') => {
                if let Some(entry) = self.selected_entry() {
                    match entry.kind {
                        FileBrowserEntryKind::Current => {
                            return FileBrowserOutcome::Selected(relative_display(
                                &self.current_dir,
                                &self.workspace_root,
                            ));
                        }
                        FileBrowserEntryKind::Parent => {
                            self.go_up();
                            return FileBrowserOutcome::None;
                        }
                        FileBrowserEntryKind::Directory => {
                            return FileBrowserOutcome::Selected(relative_display(
                                &entry.path,
                                &self.workspace_root,
                            ));
                        }
                        FileBrowserEntryKind::Manifest => {
                            let target = entry
                                .path
                                .parent()
                                .map(Path::to_path_buf)
                                .unwrap_or_else(|| self.current_dir.clone());
                            return FileBrowserOutcome::Selected(relative_display(
                                &target,
                                &self.workspace_root,
                            ));
                        }
                    }
                }
                FileBrowserOutcome::None
            }
            KeyCode::Left | KeyCode::Backspace | KeyCode::Char('h') => {
                self.go_up();
                FileBrowserOutcome::None
            }
            KeyCode::Right | KeyCode::Char('l') | KeyCode::Char('/') => {
                if let Some(entry) = self.selected_entry()
                    && matches!(entry.kind, FileBrowserEntryKind::Directory)
                {
                    self.descend(entry.path.clone());
                }
                FileBrowserOutcome::None
            }
            _ => FileBrowserOutcome::None,
        }
    }

    fn visible_rows(&self) -> usize {
        self.visible_rows.get().max(1)
    }

    fn selected_entry(&self) -> Option<&FileBrowserEntry> {
        self.scroll
            .selected_idx
            .and_then(|idx| self.entries.get(idx))
    }

    fn go_up(&mut self) {
        if self.current_dir == self.workspace_root {
            return;
        }
        if let Some(parent) = self.current_dir.parent() {
            self.current_dir = parent.to_path_buf();
            self.refresh_entries();
        }
    }

    fn descend(&mut self, candidate: PathBuf) {
        if !candidate.starts_with(&self.workspace_root) {
            self.error = Some("Cannot leave workspace root".to_string());
            return;
        }
        if candidate.is_dir() {
            self.current_dir = candidate;
            self.refresh_entries();
        }
    }

    fn refresh_entries(&mut self) {
        self.entries.clear();
        self.scroll = ScrollState::new();

        let mut items: Vec<FileBrowserEntry> = Vec::new();
        items.push(FileBrowserEntry {
            label: format!("üìÅ {}", "Use this directory"),
            summary: Some(relative_display(&self.current_dir, &self.workspace_root)),
            path: self.current_dir.clone(),
            kind: FileBrowserEntryKind::Current,
        });

        if self.current_dir != self.workspace_root {
            if let Some(parent) = self.current_dir.parent() {
                items.push(FileBrowserEntry {
                    label: "‚¨Ü  Parent".to_string(),
                    summary: Some(relative_display(parent, &self.workspace_root)),
                    path: parent.to_path_buf(),
                    kind: FileBrowserEntryKind::Parent,
                });
            }
        }

        match fs::read_dir(&self.current_dir) {
            Ok(read_dir) => {
                let mut dirs = Vec::new();
                let mut manifests = Vec::new();
                for entry in read_dir.flatten() {
                    let path = entry.path();
                    let file_name = entry.file_name().to_string_lossy().into_owned();
                    if let Ok(ft) = entry.file_type() {
                        if ft.is_dir() {
                            dirs.push(FileBrowserEntry {
                                label: format!("üìÅ {file_name}"),
                                summary: None,
                                path,
                                kind: FileBrowserEntryKind::Directory,
                            });
                        } else if ft.is_file() && looks_like_manifest(&file_name) {
                            manifests.push(FileBrowserEntry {
                                label: format!("üìÑ {file_name}"),
                                summary: Some("MCP manifest".to_string()),
                                path,
                                kind: FileBrowserEntryKind::Manifest,
                            });
                        }
                    }
                }
                dirs.sort_by(|a, b| a.label.cmp(&b.label));
                manifests.sort_by(|a, b| compare_manifest_entries(a, b));
                items.extend(dirs);
                items.extend(manifests);
            }
            Err(err) => {
                self.error = Some(format!("Failed to read directory: {err}"));
            }
        }

        self.entries = items;
        self.scroll.clamp_selection(self.entries.len());
    }
}

fn normalize_start(start: &Path, root: &Path) -> PathBuf {
    if start.is_dir() && start.starts_with(root) {
        return start.to_path_buf();
    }
    root.to_path_buf()
}

fn relative_display(path: &Path, root: &Path) -> String {
    match pathdiff::diff_paths(path, root) {
        Some(rel) if !rel.as_os_str().is_empty() => rel.to_string_lossy().into_owned(),
        _ => ".".to_string(),
    }
}

fn looks_like_manifest(name: &str) -> bool {
    matches!(
        name,
        "mcp.json" | "server.toml" | "mcp.toml" | "codex-mcp.toml"
    )
}

fn compare_manifest_entries(a: &FileBrowserEntry, b: &FileBrowserEntry) -> Ordering {
    a.label.cmp(&b.label)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn relative_display_for_root_is_dot() {
        let root = PathBuf::from("/tmp/work");
        assert_eq!(relative_display(&root, &root), ".");
    }
}
