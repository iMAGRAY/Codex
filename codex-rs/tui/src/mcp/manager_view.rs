use std::cell::Cell;
use std::path::PathBuf;

use codex_core::mcp::health::HealthReport;
use codex_core::mcp::health::HealthStatus;
use ratatui::buffer::Buffer;
use ratatui::layout::Constraint;
use ratatui::layout::Layout;
use ratatui::layout::Rect;
use ratatui::style::Modifier;
use ratatui::style::Style;
use ratatui::style::Stylize;
use ratatui::text::Line;
use ratatui::text::Span;
use ratatui::widgets::Block;
use ratatui::widgets::Borders;
use ratatui::widgets::Paragraph;
use ratatui::widgets::Row;
use ratatui::widgets::StatefulWidget;
use ratatui::widgets::Table;
use ratatui::widgets::TableState;
use ratatui::widgets::Widget;
use ratatui::widgets::Wrap;

use super::file_browser::FileBrowser;
use super::file_browser::FileBrowserOutcome;
use crate::app_event::AppEvent;
use crate::app_event_sender::AppEventSender;
use crate::bottom_pane::BottomPane;
use crate::bottom_pane::BottomPaneView;
use crate::bottom_pane::CancellationEvent;
use crate::bottom_pane::ScrollState;
use crate::mcp::types::McpServerSnapshot;
use crate::mcp::types::McpWizardDraft;

#[derive(Clone)]
pub(crate) struct McpManagerEntry {
    pub snapshot: McpServerSnapshot,
    pub health: HealthReport,
}

pub(crate) struct McpManagerInit {
    pub app_event_tx: AppEventSender,
    pub entries: Vec<McpManagerEntry>,
    pub template_count: usize,
    pub workspace_root: PathBuf,
}

pub(crate) struct McpManagerView {
    app_event_tx: AppEventSender,
    entries: Vec<McpManagerEntry>,
    template_count: usize,
    scroll: ScrollState,
    list_visible_rows: Cell<usize>,
    close_requested: bool,
    confirm_delete: bool,
    workspace_root: PathBuf,
    file_browser: Option<FileBrowser>,
}

impl McpManagerView {
    pub(crate) fn new(init: McpManagerInit) -> Self {
        let mut scroll = ScrollState::new();
        let total_rows = init.entries.len().saturating_add(1);
        scroll.clamp_selection(total_rows);
        Self {
            app_event_tx: init.app_event_tx,
            entries: init.entries,
            template_count: init.template_count,
            scroll,
            list_visible_rows: Cell::new(1),
            close_requested: false,
            confirm_delete: false,
            workspace_root: init.workspace_root,
            file_browser: None,
        }
    }

    fn selected_entry(&self) -> Option<&McpManagerEntry> {
        self.scroll
            .selected_idx
            .and_then(|idx| idx.checked_sub(1))
            .and_then(|idx| self.entries.get(idx))
    }

    fn visible_rows(&self) -> usize {
        self.list_visible_rows.get().max(1)
    }

    fn total_rows(&self) -> usize {
        self.entries.len().saturating_add(1)
    }

    fn move_selection_up(&mut self) {
        self.scroll.move_up_wrap(self.total_rows());
        self.scroll
            .ensure_visible(self.total_rows(), self.visible_rows());
        self.confirm_delete = false;
    }

    fn move_selection_down(&mut self) {
        self.scroll.move_down_wrap(self.total_rows());
        self.scroll
            .ensure_visible(self.total_rows(), self.visible_rows());
        self.confirm_delete = false;
    }

    fn page_up(&mut self) {
        let steps = self.visible_rows().saturating_sub(1);
        for _ in 0..steps {
            self.scroll.move_up_wrap(self.total_rows());
        }
        self.scroll
            .ensure_visible(self.total_rows(), self.visible_rows());
        self.confirm_delete = false;
    }

    fn page_down(&mut self) {
        let steps = self.visible_rows().saturating_sub(1);
        for _ in 0..steps {
            self.scroll.move_down_wrap(self.total_rows());
        }
        self.scroll
            .ensure_visible(self.total_rows(), self.visible_rows());
        self.confirm_delete = false;
    }

    fn open_new_wizard(&self) {
        self.app_event_tx.send(AppEvent::OpenMcpWizard {
            template_id: None,
            draft: None,
            existing_name: None,
        });
    }

    fn open_edit_wizard(&self) {
        if let Some(entry) = self.selected_entry() {
            let config = entry.snapshot.to_config();
            let draft = McpWizardDraft::from_existing(entry.snapshot.name.clone(), &config);
            self.app_event_tx.send(AppEvent::OpenMcpWizard {
                template_id: entry.snapshot.template_id.clone(),
                draft: Some(draft),
                existing_name: Some(entry.snapshot.name.clone()),
            });
        }
    }

    fn request_remove_selected(&mut self) {
        if self.selected_entry().is_none() {
            self.confirm_delete = false;
            return;
        }
        if self.confirm_delete {
            if let Some(entry) = self.selected_entry() {
                self.app_event_tx.send(AppEvent::RemoveMcpServer {
                    name: entry.snapshot.name.clone(),
                });
            }
            self.confirm_delete = false;
        } else {
            self.confirm_delete = true;
        }
    }

    fn refresh(&self) {
        self.app_event_tx.send(AppEvent::ReloadMcpServers);
    }

    fn open_file_browser(&mut self) {
        let start = self.workspace_root.clone();
        self.file_browser = Some(FileBrowser::new(start, self.workspace_root.clone()));
        self.confirm_delete = false;
    }

    fn handle_file_browser_key(&mut self, key_event: crossterm::event::KeyEvent) {
        if let Some(browser) = self.file_browser.as_mut() {
            match browser.handle_key(key_event) {
                FileBrowserOutcome::None => {}
                FileBrowserOutcome::Cancelled => {
                    self.file_browser = None;
                }
                FileBrowserOutcome::Selected(source_path) => {
                    let mut draft = McpWizardDraft::default();
                    draft.source_path = source_path;
                    self.app_event_tx.send(AppEvent::OpenMcpWizard {
                        template_id: None,
                        draft: Some(draft),
                        existing_name: None,
                    });
                    self.file_browser = None;
                }
            }
        }
    }

    fn render_list(&self, area: Rect, buf: &mut Buffer) {
        let header = Row::new(vec!["Name".bold(), "Status".bold(), "Command".bold()]);

        let mut rows: Vec<Row> = Vec::with_capacity(self.entries.len().saturating_add(1));
        rows.push(Row::new(vec![
            Span::styled("Browse workspace…", Style::default().italic()),
            "".into(),
            "Press Enter to select a manifest".dim().into(),
        ]));

        rows.extend(self.entries.iter().map(|entry| {
            Row::new(vec![
                Span::raw(entry.snapshot.name.clone()),
                health_span(&entry.health.status),
                Span::raw(entry.snapshot.command.clone()),
            ])
        }));

        let mut state = TableState::default();
        state.select(self.scroll.selected_idx);
        *state.offset_mut() = self.scroll.scroll_top;

        let table = Table::new(
            rows,
            [
                Constraint::Percentage(28),
                Constraint::Length(10),
                Constraint::Percentage(62),
            ],
        )
        .header(header)
        .block(
            Block::default()
                .borders(Borders::ALL)
                .title("Configured MCP Servers"),
        )
        .row_highlight_style(Style::default().add_modifier(Modifier::REVERSED));

        StatefulWidget::render(table, area, buf, &mut state);

        let visible_rows = area.height.saturating_sub(3) as usize; // header + borders
        self.list_visible_rows.set(visible_rows.max(1));
    }

    fn render_detail(&self, area: Rect, buf: &mut Buffer) {
        let mut lines: Vec<Line> = Vec::new();
        if self.scroll.selected_idx == Some(0) {
            lines.push("Browse the workspace to add MCP servers.".into());
            lines.push(
                "Press Enter to open the file browser or 'b' to browse directly."
                    .dim()
                    .into(),
            );
            lines.push(
                "Codex will analyse the selected folder and pre-fill the wizard."
                    .dim()
                    .into(),
            );
        } else if let Some(entry) = self.selected_entry() {
            lines.extend(detail_lines(&entry.snapshot));
            lines.push(Line::from(vec![
                "Status: ".dim(),
                health_span(&entry.health.status),
            ]));
            if let Some(notes) = entry.health.notes.as_ref() {
                lines.push(Line::from(vec!["  • ".into(), notes.clone().dim()]));
            }
            if self.confirm_delete {
                lines.push(Line::from(""));
                lines.push("Press 'd' again to confirm deletion".red().into());
            }
        } else if self.entries.is_empty() {
            lines.push("No MCP servers configured.".into());
            if self.template_count > 0 {
                lines.push(
                    format!(
                        "Press 'n' to create one using {} available template(s).",
                        self.template_count
                    )
                    .dim()
                    .into(),
                );
            }
            lines.push(
                "Press 'f' to browse the workspace for MCP manifests."
                    .dim()
                    .into(),
            );
        }

        lines.push(Line::from(""));
        lines.push(Line::from(vec![
            "Keys: ".dim(),
            "↑/↓".cyan(),
            " move  ".dim(),
            "n".cyan(),
            " new  ".dim(),
            "f/b".cyan(),
            " browse  ".dim(),
            "Enter".cyan(),
            " select/edit  ".dim(),
            "r".cyan(),
            " reload  ".dim(),
            "d".cyan(),
            " delete  ".dim(),
            "Esc".cyan(),
            " close".dim(),
        ]));

        let paragraph = Paragraph::new(lines)
            .block(Block::default().borders(Borders::ALL).title("Details"))
            .wrap(Wrap { trim: true });
        Widget::render(paragraph, area, buf);
    }
}

impl BottomPaneView for McpManagerView {
    fn handle_key_event(&mut self, _pane: &mut BottomPane, key_event: crossterm::event::KeyEvent) {
        use crossterm::event::KeyCode;

        if self.file_browser.is_some() {
            self.handle_file_browser_key(key_event);
            return;
        }

        self.confirm_delete &= matches!(key_event.code, KeyCode::Char('d'));

        match key_event.code {
            KeyCode::Esc | KeyCode::Char('q') => {
                self.close_requested = true;
            }
            KeyCode::Char('n') => self.open_new_wizard(),
            KeyCode::Char('f') | KeyCode::Char('b') => self.open_file_browser(),
            KeyCode::Enter | KeyCode::Char('e') => {
                if self.scroll.selected_idx == Some(0) {
                    self.open_file_browser();
                } else {
                    self.open_edit_wizard();
                }
            }
            KeyCode::Char('r') => self.refresh(),
            KeyCode::Char('d') => self.request_remove_selected(),
            KeyCode::Up => self.move_selection_up(),
            KeyCode::Down => self.move_selection_down(),
            KeyCode::PageUp => self.page_up(),
            KeyCode::PageDown => self.page_down(),
            _ => {
                if !matches!(key_event.code, KeyCode::Char('d')) {
                    self.confirm_delete = false;
                }
            }
        }
    }

    fn is_complete(&self) -> bool {
        self.close_requested
    }

    fn on_ctrl_c(&mut self, _pane: &mut BottomPane) -> CancellationEvent {
        self.close_requested = true;
        CancellationEvent::Handled
    }

    fn desired_height(&self, _width: u16) -> u16 {
        if self.file_browser.is_some() { 28 } else { 22 }
    }

    fn render(&self, area: Rect, buf: &mut Buffer) {
        if let Some(browser) = self.file_browser.as_ref() {
            browser.render(area, buf, "Select MCP Source");
            return;
        }

        let layout = Layout::horizontal([Constraint::Percentage(46), Constraint::Percentage(54)])
            .split(area);
        self.render_list(layout[0], buf);
        self.render_detail(layout[1], buf);
    }
}

fn health_span(status: &HealthStatus) -> Span<'static> {
    match status {
        HealthStatus::Unknown => "unknown".dim(),
        HealthStatus::Passing => "passing".green(),
        HealthStatus::Failing => "failing".red(),
    }
}

fn detail_lines(snapshot: &McpServerSnapshot) -> Vec<Line<'static>> {
    let mut lines = Vec::new();
    lines.push(Line::from(vec![
        "Name: ".dim(),
        snapshot.name.clone().into(),
        "  Command: ".dim(),
        snapshot.command.clone().into(),
    ]));

    if !snapshot.args.is_empty() {
        lines.push(Line::from(vec![
            "Args: ".dim(),
            snapshot.args.join(" ").into(),
        ]));
    }

    if let Some(desc) = snapshot.description.as_ref() {
        lines.push(Line::from(vec!["Description: ".dim(), desc.clone().into()]));
    }

    if !snapshot.tags.is_empty() {
        lines.push(Line::from(vec![
            "Tags: ".dim(),
            snapshot.tags.join(", ").into(),
        ]));
    }

    if let Some(template_id) = snapshot.template_id.as_ref() {
        lines.push(Line::from(vec![
            "Template: ".dim(),
            template_id.clone().into(),
        ]));
    }

    if let Some(timeout) = snapshot.startup_timeout_ms {
        lines.push(Line::from(vec![
            "Startup timeout (ms): ".dim(),
            timeout.to_string().into(),
        ]));
    }

    if !snapshot.env.is_empty() {
        lines.push("Env:".dim().into());
        for (idx, (k, v)) in snapshot.env.iter().enumerate() {
            if idx >= 6 {
                lines.push(
                    format!("  … {} more", snapshot.env.len() - idx)
                        .dim()
                        .into(),
                );
                break;
            }
            lines.push(Line::from(vec!["  • ".into(), format!("{k}={v}").into()]));
        }
    }

    if let Some(auth) = snapshot.auth.as_ref() {
        lines.extend(auth_lines(auth));
    }

    if let Some(health) = snapshot.health.as_ref() {
        lines.extend(health_lines(health));
    }

    lines
}

fn auth_lines(auth: &crate::mcp::types::AuthDraft) -> Vec<Line<'static>> {
    let mut lines = Vec::new();
    lines.push("Auth:".dim().into());
    if let Some(kind) = auth.kind.as_ref() {
        lines.push(Line::from(vec!["  • Type: ".into(), kind.clone().into()]));
    }
    if let Some(secret) = auth.secret_ref.as_ref() {
        lines.push(Line::from(vec![
            "  • Secret: ".into(),
            secret.clone().into(),
        ]));
    }
    if !auth.env.is_empty() {
        for (idx, (k, v)) in auth.env.iter().enumerate() {
            if idx >= 4 {
                lines.push(format!("  … {} more", auth.env.len() - idx).dim().into());
                break;
            }
            lines.push(Line::from(vec![
                "     - ".into(),
                format!("{k}={v}").into(),
            ]));
        }
    }
    lines
}

fn health_lines(health: &crate::mcp::types::HealthDraft) -> Vec<Line<'static>> {
    let mut lines = Vec::new();
    lines.push("Health:".dim().into());
    if let Some(kind) = health.kind.as_ref() {
        lines.push(Line::from(vec!["  • Type: ".into(), kind.clone().into()]));
    }
    if let Some(cmd) = health.command.as_ref() {
        lines.push(Line::from(vec!["  • Command: ".into(), cmd.clone().into()]));
    }
    if !health.args.is_empty() {
        lines.push(Line::from(vec![
            "  • Args: ".into(),
            health.args.join(" ").into(),
        ]));
    }
    if let Some(endpoint) = health.endpoint.as_ref() {
        lines.push(Line::from(vec![
            "  • Endpoint: ".into(),
            endpoint.clone().into(),
        ]));
    }
    if let Some(timeout) = health.timeout_ms {
        lines.push(Line::from(vec![
            "  • Timeout (ms): ".into(),
            timeout.to_string().into(),
        ]));
    }
    if let Some(interval) = health.interval_seconds {
        lines.push(Line::from(vec![
            "  • Interval (s): ".into(),
            interval.to_string().into(),
        ]));
    }
    lines
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::mcp::types::AuthDraft;
    use crate::mcp::types::HealthDraft;
    use insta::assert_snapshot;
    use ratatui::buffer::Buffer;
    use ratatui::layout::Rect;
    use std::collections::BTreeMap;
    use tokio::sync::mpsc::unbounded_channel;

    fn render_view(view: &McpManagerView) -> String {
        let width = 72;
        let height = BottomPaneView::desired_height(view, width);
        let area = Rect::new(0, 0, width, height);
        let mut buf = Buffer::empty(area);
        BottomPaneView::render(view, area, &mut buf);

        (0..area.height)
            .map(|row| {
                let mut line = String::new();
                for col in 0..area.width {
                    let symbol = buf[(area.x + col, area.y + row)].symbol();
                    if symbol.is_empty() {
                        line.push(' ');
                    } else {
                        line.push_str(symbol);
                    }
                }
                line.trim_end().to_string()
            })
            .collect::<Vec<_>>()
            .join("\n")
    }

    fn sample_entry(name: &str, command: &str, status: HealthStatus) -> McpManagerEntry {
        let mut env = BTreeMap::new();
        env.insert("TOKEN".to_string(), "***".to_string());

        let snapshot = McpServerSnapshot {
            name: name.to_string(),
            command: command.to_string(),
            args: vec!["--serve".to_string()],
            env,
            description: Some("Sample server".to_string()),
            tags: vec!["beta".to_string(), "internal".to_string()],
            template_id: Some("sample-template".to_string()),
            auth: Some(AuthDraft {
                kind: Some("env".to_string()),
                secret_ref: Some("vault:secret".to_string()),
                env: BTreeMap::from([("AUTH_TOKEN".to_string(), "secret".to_string())]),
            }),
            health: Some(HealthDraft {
                kind: Some("stdio".to_string()),
                command: Some("health-check".to_string()),
                args: vec!["--ping".to_string()],
                timeout_ms: Some(5_000),
                interval_seconds: Some(60),
                endpoint: None,
                protocol: None,
            }),
            display_name: Some(format!("{name} MCP")),
            category: Some("general".to_string()),
            metadata: None,
            startup_timeout_ms: Some(15_000),
        };

        McpManagerEntry {
            snapshot,
            health: HealthReport::new(status).with_notes("synthetic status for test rendering"),
        }
    }

    fn make_view(entries: Vec<McpManagerEntry>, template_count: usize) -> McpManagerView {
        let (tx, _rx) = unbounded_channel();
        McpManagerView::new(McpManagerInit {
            app_event_tx: AppEventSender::new(tx),
            entries,
            template_count,
            workspace_root: PathBuf::from("/workspace"),
        })
    }

    #[test]
    fn renders_manager_with_entries() {
        let view = make_view(
            vec![
                sample_entry("anthropic", "anthropic-mcp", HealthStatus::Unknown),
                sample_entry("openai", "openai-mcp", HealthStatus::Passing),
            ],
            3,
        );

        assert_snapshot!("mcp_manager_with_entries", render_view(&view));
    }

    #[test]
    fn renders_delete_confirmation() {
        let mut view = make_view(
            vec![sample_entry(
                "perplexity",
                "perplexity-mcp",
                HealthStatus::Failing,
            )],
            1,
        );
        view.scroll.selected_idx = Some(0);
        view.confirm_delete = true;

        assert_snapshot!("mcp_manager_delete_confirm", render_view(&view));
    }

    #[test]
    fn renders_empty_state_with_template_hint() {
        let view = make_view(Vec::new(), 5);

        assert_snapshot!("mcp_manager_empty", render_view(&view));
    }
}
