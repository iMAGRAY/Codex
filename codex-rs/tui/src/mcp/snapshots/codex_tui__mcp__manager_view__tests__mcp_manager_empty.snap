---
source: tui/src/mcp/manager_view.rs
expression: render_view(&view)
---
┌Configured MCP Servers─────────┐┌Details──────────────────────────────┐
│Name      Status     Command   ││Browse the workspace to add MCP      │
│Browse wo            Press Ente││servers.                             │
│                               ││Press Enter to open the file browser │
│                               ││or 'b' to browse directly.           │
│                               ││Codex will analyse the selected      │
│                               ││folder and pre-fill the wizard.      │
│                               ││                                     │
│                               ││Keys: ↑/↓ move  n new  f/b browse    │
│                               ││Enter select/edit  r reload  d delete│
│                               ││Esc close                            │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
│                               ││                                     │
└───────────────────────────────┘└─────────────────────────────────────┘
